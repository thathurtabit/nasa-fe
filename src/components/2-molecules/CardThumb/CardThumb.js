import React, { Suspense, lazy } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import 'intersection-observer';
import { InView } from 'react-intersection-observer';
import CardThumbStyled, { CardLink } from './CardThumb.styled';
import LoadingSmall from '../../1-atoms/LoadingSmall/LoadingSmall';
import { toggleModal } from '../../../state/actions/toggleModal';

const LazyThumb = lazy(() => import('../../1-atoms/CardImage/CardImage'));

const mapStateToProps = state => ({
  modalOpen: state.modalOpen,
});

const mapDispatchToProps = dispatch => ({
  toggleModal: bool => dispatch(toggleModal(bool)),
});

const CardThumb = ({ product, modalOpen, toggleModal, num }) => {
  const { title, imgSrc, id, productNo } = product;
  const body = document.querySelector('body');

  // Breaking out of React
  if (modalOpen) {
    body.setAttribute('style', 'overflow-y: hidden;');
  } else {
    body.setAttribute('style', 'overflow-y: scroll;');
  }

  return (
    <InView>
      {({ inView, ref }) => (
        <CardThumbStyled ref={ref}>
          {inView && (
            <Suspense key={id} fallback={<LoadingSmall />}>
              <CardLink
                id={`card-thumb-${num}`}
                key={id}
                to={{
                  pathname: `/card/${productNo}`,
                  state: { modal: true },
                }}
                tabIndex="0"
                title={title}
                onClick={() => toggleModal(true)}
              >
                <LazyThumb url={imgSrc} title={title} thumb />
              </CardLink>
            </Suspense>
          )}
        </CardThumbStyled>
      )}
    </InView>
  );
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CardThumb);

CardThumb.propTypes = {
  modalOpen: PropTypes.bool,
  toggleModal: PropTypes.func.isRequired,
  product: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  ).isRequired,
};

CardThumb.defaultProps = {
  modalOpen: false,
};
