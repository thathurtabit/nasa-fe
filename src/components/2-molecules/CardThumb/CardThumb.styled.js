import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const CardLink = styled(Link)`
  position: relative;
`;

const CardThumbStyled = styled.li`
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  min-height: 150px;
  position: relative;
`;

export default CardThumbStyled;
