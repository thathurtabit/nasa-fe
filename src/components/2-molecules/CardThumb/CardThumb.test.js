import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import 'jest-styled-components';
import { shallowWithTheme } from '../../../utils/test_config/testHelpers';

import CardThumb from './CardThumb';

describe('CardThumb', () => {
  it('should render correctly', () => {
    const element = shallowWithTheme(
      <CardThumb card={{ Title: 'Card Title' }} product={{}} />
    );
    expect(shallowToJson(element)).toMatchSnapshot();
  });
});
