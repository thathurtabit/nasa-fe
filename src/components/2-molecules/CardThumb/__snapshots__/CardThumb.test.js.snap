// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardThumb should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
